enterprise/server/iprules/iprules.go

```diff
@@ -0,0 +1,261 @@
+		return &noopIpRuleCache{}, nil
+	}
+	config := &lru.Config[*ipRuleCacheEntry]{
+		MaxSize: cacheSize,
+		SizeFn:  func(v *ipRuleCacheEntry) int64 { return int64(len(v.allowed)) },
+	}
+	l, err := lru.NewLRU[*ipRuleCacheEntry](config)
+	if err != nil {
+		return nil, err
+	}
+	return &memIpRuleCache{
+		lru: l,
+	}, nil
+}
+
+type Service struct {
+	env environment.Env
+
+	cache ipRuleCache
+}
+
+func New(env environment.Env) (*Service, error) {
+	cache, err := newIpRuleCache()
+	if err != nil {
+		return nil, err
+	}
+	return &Service{
+		env:   env,
+		cache: cache,
+	}, nil
+}
+
+func Register(env environment.Env) error {
+	if !*enableIPRules {
+		return nil
+	}
+
+	enforcer, err := New(env)
+	if err != nil {
+		return err
+	}
+	env.SetIPRulesService(enforcer)
+	return nil
+}
+
+func (s *Service) loadRulesFromDB(ctx context.Context, groupID string) ([]*net.IPNet, error) {
+	rows, err := s.env.GetDBHandle().DB(ctx).Raw(
+		`SELECT * FROM "IPRules" WHERE group_id = ?`, groupID).Rows()
+	if err != nil {
+		return nil, err
+	}
+
+	var allowed []*net.IPNet
+	for rows.Next() {
+		r := &tables.IPRule{}
+		if err := s.env.GetDBHandle().DB(ctx).ScanRows(rows, r); err != nil {
+			return nil, err
+		}
+		_, ipNet, err := net.ParseCIDR(r.CIDR)
+		if err != nil {
+			alert.UnexpectedEvent("unparsable CIDR rule", "rule %q", r.CIDR)
+			continue
+		}
+		allowed = append(allowed, ipNet)
+	}
+	return allowed, nil
+}
+
+func (s *Service) AuthorizeGroup(ctx context.Context, groupID string) error {
+	g, err := s.env.GetUserDB().GetGroupByID(ctx, groupID)
+	if err != nil {
+		return err
+	}
+	if !g.EnforceIPRules {
+		return nil
+	}
+
+	rawClientIP := clientip.Get(ctx)
+	clientIP := net.ParseIP(rawClientIP)
+	// Client IP is not parsable.
+	if clientIP == nil {
+		return status.FailedPreconditionErrorf("client IP %q is not valid", rawClientIP)
+	}
+
+	allowed, ok := s.cache.Get(groupID)
+	if !ok {
+		rs, err := s.loadRulesFromDB(ctx, groupID)
+		if err != nil {
+			return err
+		}
+		s.cache.Add(groupID, rs)
+		allowed = rs
+	}
+
+	for _, a := range allowed {
+		if a.Contains(clientIP) {
+			return nil
+		}
+	}
+
+	return status.PermissionDeniedErrorf("Client %q is not allowed by Organization IP rules", rawClientIP)
+}
+
+func (s *Service) Authorize(ctx context.Context) error {
+	u, err := perms.AuthenticatedUser(ctx, s.env)
+	if authutil.IsAnonymousUserError(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// Server admins in impersonation mode can bypass IP rules.
+	if u.IsImpersonating() {
+		return nil
+	}
+
+	if !u.GetEnforceIPRules() {
+		return nil
+	}
+
+	return s.AuthorizeGroup(ctx, u.GetGroupID())
+}
+
+func (s *Service) AddRule(ctx context.Context, req *irpb.AddRuleRequest) (*irpb.AddRuleResponse, error) {
+	u, err := perms.AuthenticatedUser(ctx, s.env)
+	if err != nil {
+		return nil, err
+	}
+	if err := authutil.AuthorizeGroupRole(u, req.GetRequestContext().GetGroupId(), role.Admin); err != nil {
+		return nil, err
+	}
```

Comment:
could you have a 

`func (s *Service) checkPreconditions(ctx, reqCtx) error` method that does this since it's the same check in all 3 places? (Add|Update|Delete)